### ISSUE TITLE
Raycasting of individual meshes is broken

### ISSUE THREAD

**Issue by liberostelios** [NASA-AMMOS/3DTilesRendererJS#168]
The current implementation of raycasting for tiles breaks the traditional functionality of `Mesh.raycast()`. I believe this is a risky move because the user of the library doesn't know that and if they call `.intersectObjects()` directly on the meshes they get no return.

The current solution is very efficient and nicely thought, but this is a very important side-effect. I admit that I couldn't find a better solution to this. Maybe there is a way that `Mesh.raycast()` is only set to `emptyRaycast()` during the traversal of the tree?

Otherwise, I think it should be at least mention it in the docs as a warning. And maybe also mention that a workaround to the current situation is to loop through the meshes calling the prototype function instead:

```
let result = [];

Object.getPrototypeOf( mesh ).raycast.call( mesh, raycaster, result );
```

**Comment by gkjohnson** [comment]
Hey @liberostelios! Yes there have been a few trades related to tiles made for the sake of performance including disabling raycasting on the them so the bounding volume hierarchy can be traversed in the group.raycast function. It's also not possible to reposition the tiles with `position`, `rotation`, etc without calling `updateMatrixWorld` manually because [automatic matrix updates are disabled unless the group moves](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/7ad0e30d2c9eaf352842a502678835b720452bb7/src/three/TilesGroup.js#L22-L80) since the child tiles are not expected to move.

I agree it's worth documenting both of these things. I'm not exactly sure where, though? Maybe start in the TilesRenderer class section? Or the `TilesGroup` object could be documented explicitly with some of these caveats.

> maybe also mention that a workaround to the current situation is to loop through the meshes calling the prototype function instead

What's your use case for raycasting against the individual tiles? My expectation was that users wouldn't / shouldn't need to know about individual tiles because the renderer needs explicit control over updating them and adding and removing them from the scene. The terrain should be treated as a whole group. If there's a compelling need, though, then maybe a toggle could be added to enable or disable the optimized raycast functionality.

> Maybe there is a way that Mesh.raycast() is only set to emptyRaycast() during the traversal of the tree?

This is an interesting idea but unfortunately there's no way to know when a raycast is set to `traverse = true` or when the children have finished being iterated over, as far as I know. But I'm open to ideas!

**Comment by liberostelios** [comment]
Hey @gkjohnson! Thanks for the quick response.

> I agree it's worth documenting both of these things. I'm not exactly sure where, though? Maybe start in the TilesRenderer class section? Or the `TilesGroup` object could be documented explicitly with some of these caveats.

I don't have a strong opinion here. I think an example under the *Use* section with clarification about the overrides of typical threejs behaviour would suffice.

> What's your use case for raycasting against the individual tiles? My expectation was that users wouldn't / shouldn't need to know about individual tiles because the renderer needs explicit control over updating them and adding and removing them from the scene. The terrain should be treated as a whole group. If there's a compelling need, though, then maybe a toggle could be added to enable or disable the optimized raycast functionality.

I know in our case we assumed that all threejs objects keep their functionality. We had a problem with object picking which was caused by our tileset being slightly off, but we didn't realise until we we figured out that TilesRenderer overrides threejs behaviour. Part of the process was to validate that the geometries are fine, so we built a simple array with all the meshes and called `intersectObjects()` on them with no return. We were confused here because we assumed that as soon as the tiles are added in the scene they are just "dumb" threejs objects, which was obviously not the case. In our minds it was like "the tileset can't be the problem, because raycasting has nothing to do with it". 

Actually, adding an option is probably the best idea, because not only this offers an option to revert the objects' original threejs behaviour (probably not super important, but still nice to have), but the presence of the option itself informs the user that there is something special about how the library treats its meshes.

> This is an interesting idea but unfortunately there's no way to know when a raycast is set to `traverse = true` or when the children have finished being iterated over, as far as I know. But I'm open to ideas!

You are right, there doesn't seem to be a way to know when threejs finishes the traversal. One option, though, would be to offer the existing functionality as an individual function (e.g. `.raycastTiles()`) and let the original behaviour be the "dumb" raycasting that threejs does.

So, to sum up, I can see three possibilities:

1. Add an example of raycasting and explain what's happening (and how people can override the behaviour).
2. Add an option to disable the overridden behaviour.
3. Extract the current behaviour to its own function and let threejs's `raycast()` work in its own way.

If you ask me, I am happy with the first option and it's the least time consuming. But probably the best one is the second.

**Comment by liberostelios** [comment]
> This is an interesting idea but unfortunately there's no way to know when a raycast is set to `traverse = true` or when the children have finished being iterated over, as far as I know. But I'm open to ideas!

Okay, I just realised that we know when the children are visited because `emptyRaycast()` is been called. So this is doable by making `emptyRaycast()` set `this.raycast` back to its prototype function. That's possible, right? I am just not sure if there is any asynchronous logic here which can break the assumption that `emptyRaycast` will always be called for the same group's `raycast()` iteration.

**Comment by gkjohnson** [comment]
> Extract the current behaviour to its own function and let threejs's raycast() work in its own way.

There actually already is a [function on TilesRenderer](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/src/three/TilesRenderer.js#L173) for raycasting already it's just not documented and is used by the TilesGroup raycast.

> Okay, I just realised that we know when the children are visited because emptyRaycast() is been called. So this is doable by making emptyRaycast() set this.raycast back to its prototype function. That's possible, right?

I don't believe there's a way to know if the the child tiles are going to be traversed or not so I think this approach will wind up causing some confusing situations, as well.

I think I'm inclined to make change number 2 and add a variable on TilesRenderer to enable the optimized raycasting that defaults to false so it must be opted in to to avoid issues like this one again. I'm trying to think of what it should be called, though. Maybe `TilesRenderer.optimizedRaycast`? `overrideRaycast`? `enableBoundsRaycast`? Any other ideas?

Are you interested in adding the feature? I figure a custom raycast override function can be created per TilesRenderer and if the optimized raycast is disabled then the raycast override function can call `Object.getPrototypeOf( this ).raycast.call( this, ...args )`.

**Comment by liberostelios** [comment]
> There actually already is a [function on TilesRenderer](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/src/three/TilesRenderer.js#L173) for raycasting already it's just not documented and is used by the TilesGroup raycast.

What I meant is that the existing `raycast()` function could be called something else (e.g. `raycastTiles()`) so when a user calls the typical raycasting over the group it will be handled entirely by threejs (inefficiently) and if they want to do it efficiently they'll have to call `raycastTiles()` explicitly.

> I don't believe there's a way to know if the the child tiles are going to be traversed or not so I think this approach will wind up causing some confusing situations, as well.

Ah, yes. I missed that point. That's, indeed, not possible.

> I think I'm inclined to make change number 2 and add a variable on TilesRenderer to enable the optimized raycasting that defaults to false so it must be opted in to to avoid issues like this one again. I'm trying to think of what it should be called, though. Maybe `TilesRenderer.optimizedRaycast`? `overrideRaycast`? `enableBoundsRaycast`? Any other ideas?
> 
> Are you interested in adding the feature? I figure a custom raycast override function can be created per TilesRenderer and if the optimized raycast is disabled then the raycast override function can call `Object.getPrototypeOf( this ).raycast.call( this, ...args )`.

Sure! I'll work on it.

**Comment by gkjohnson** [comment]
> Sure! I'll work on it.

Awesome, thank you!

**Comment by gkjohnson** [comment]
Fixed in #169
